File: api/src/services/events.py
import json

async def publish_event(db, event_type: str, payload: dict, basket_id: str = None, workspace_id: str = None, actor_id: str = None):
    # Migrated from basket_events to canonical events table
    query = """
        INSERT INTO events (kind, payload, basket_id, workspace_id, origin, actor_id, ts)
        VALUES (:kind, :payload, :basket_id, :workspace_id, :origin, :actor_id, NOW())
    """
    await db.execute(query, {
        "kind": event_type,
        "payload": json.dumps(payload),
        "basket_id": basket_id,
        "workspace_id": workspace_id,
        "origin": "system",
        "actor_id": actor_id
    })

File: web/shared/contracts/events.ts
// Event system types (canonical events table)

export type Event = {
  id: string;
  basket_id: string;
  workspace_id: string;
  kind: string;
  payload: Record<string, any>;
  origin: "user" | "system" | "agent";
  actor_id?: string;
  agent_type?: string;
  block_id?: string;
  ts: string;
};

// Pagination types
export type EventsPage = {
  events: Event[];
  last_cursor: {
    ts: string;
    id: string;
  };
  has_more: boolean;
};